rfd/0009-viewer-app.md

```diff
@@ -0,0 +1,76 @@
+---
+authors: Mathew (wrm932@gmail.com)
+state: draft
+---
+
+# Required Approvers
+* Engineering @pierrebeaucamp && @jimbishopp && @michellescripts && @mcbattirola
+
+# RFD 0009 - RFDs
+
+## What
+
+Create a directory viewer similar to Github and Google drive. 
+
+
+## Why
+
+A pratical exam that tests the potential applicant on their fullstack software knowledge. 
+
+## Details
+
+Each RFD is stored in a markdown file under
+https://github.com/gravitational/teleport/tree/master/rfd and has a unique
+number.
+
+The directory viewer should be allow the user to navigate a directory structure by clicking on files or folders. A bread crumb navigation should inform them of where they currently are and allow them to navigate back to their previously clicked directory location. The user should also be able to filter out a list of files using a search box. Search/filtering will be done client side. 
+
+ In addition, user authentication should be implemented so only logged in users will be able to access the viewer. If a user is unauthenticated they should be redirected back to the login form. However, if given a direct url that points to a specific directory. The application should be able to remember it and redirect the unauthenticated user to the proper url after successful authentication. 
+
+## API 
+
+API should return a directory structure json response 1-level deep at a time. Subsequent calls to this api with added directory parameters should return the appropriate json response. 
+
+```json
+GET /directory/[...slug]
```

Comment:
What are the possible return codes?

Reply:
How are you going to treat folders that have non-URL-safe symbols in their names?

Reply:
Good point, possible return codes for the directory api will be
* `200` for successful response
* `401` for unauthenticated requests
* `500` for server error

For handling non-URL-safe symbols for folder names. How about we run a cleaning function on files and folder names before outputting the response.